README.markdown

```diff
@@ -392,6 +392,34 @@ You can find real world examples of `Doctest` being used below:
   * [base Data/Maybe.hs](https://github.com/ghc/ghc/blob/669cbef03c220de43b0f88f2b2238bf3c02ed64c/libraries/base/Data/Maybe.hs#L36-L79)
   * [base Data/Functor.hs](https://github.com/ghc/ghc/blob/669cbef03c220de43b0f88f2b2238bf3c02ed64c/libraries/base/Data/Functor.hs#L34-L64)
 
+## Test dependencies
+By default `doctest` runs all test in the same GHCi session. This means some
+operations, e.g. `:set -XTypeApplications`, carry over to the new module creating
+a potentially unexpected dependency between tests. In order to isolate tests,
+you can use `--isolate-modules`.
```

Comment:
Cross-module contamination should be a doctest bug anyway. Is this something peculiar to TypeApplications, or is it happening more generally with extensions and possibly even other things? Let's have a test demonstrating the necessity of fresh GHCi processes, so that no-one in the future will be tempted to 'optimise' doctest by re-using them.

I'd rather isolate modules unconditionally, because flaky tests are terrible. I presume there's a speed penalty to doing that, but what's the actual damage (assuming `--use-package-db`)?

Reply:
Agreed, this quest started by flaky behavior in the first place. I kinda overshot my goal when I saw how easy it would be to add multithreading too :-).

> Is this something peculiar to TypeApplications, or is it happening more generally with extensions and possibly even other things?

This affects anything that's not reset by `:reload`. So everything `:set`, but also imports. I can't really think of any others, but these two unintendedly make module tests very much dependent on each other in my day job's Haskell project.

> I presume there's a speed penalty to doing that, but what's the actual damage (assuming --use-package-db)?

Given `--use-package-db` there should be that much of a penalty. GHCi can start and load a module from a package db in under 100 ms on my machine:

```
martijn@deskei:~$ time bash -c 'echo "import Data.List" | ghci'
GHCi, version 8.10.3: https://www.haskell.org/ghc/  :? for help
Prelude> Prelude Data.List> Leaving GHCi.

real	0m0.087s
```

Reply:
OK - so the performance hit is reasonable, so let's have module isolation on by default, because playing whack-a-mole with things that survive `:reload` does not look like a winning game. *Maybe* an option to turn it off (or re-use the fast switch?) if it really is painful for large projects or something.

Reply:
Agreed! I do think we're [going to break a few workflows with this though](https://xkcd.com/1172/) :-), so we should probably add a flag to disable it.